src/data/list/cycle.lean

```diff
@@ -553,13 +553,35 @@ The `s : cycle α` as a `multiset α`.
 def to_multiset (s : cycle α) : multiset α :=
 quotient.lift_on' s (λ l, (l : multiset α)) (λ l₁ l₂ (h : l₁ ~r l₂), multiset.coe_eq_coe.mpr h.perm)
 
+/--
+The lift of `list.map`.
+-/
+def map {β : Type*} (f : α → β) (s : cycle α) : cycle β :=
+quot.lift_on s (λ l : list α, (l.map f : cycle β))
+  (λ l₁ l₂ (h : l₁ ~r l₂), by simpa using h.map _)
+
 section decidable
 
 variable [decidable_eq α]
 
 instance {s : cycle α} : decidable (nodup s) :=
 quot.rec_on_subsingleton s (λ (l : list α), list.nodup_decidable l)
 
+/--
+The `finset` of lists that can make the cycle.
+-/
+def lists (s : cycle α) : finset (list α) :=
+quotient.lift_on' s (λ l, (l.permutations.filter (λ (l' : list α), (l' : cycle α) = s)).to_finset) $
+  λ l₁ l₂ (h : l₁ ~r l₂),
+  begin
+    induction s using quotient.induction_on',
+    ext,
+    simp only [mem_filter, coe_eq_coe, mk'_eq_coe, and.congr_left_iff, mem_permutations,
+               mem_to_finset],
+    intro,
+    exact ⟨λ H, H.trans h.perm, λ H, H.trans h.perm.symm⟩
+  end
```

Comment:
I think there might be a clever way to prove this if you can show:
```lean
-- this is true, right?
def _root_.list.is_rotated.permutations (l₁ l₂ : list α) (h : l₁ ~r l₂) : l₁.permutations ~ l₂.permutations :=
begin
  sorry,
end
```
from which you can get
```lean
def lists' : cycle α → multiset (list α) :=
quotient.map' (λ l, list.permutations l) $λ l₁ l₂ h, h.permutations
```

Reply:
Yes, that would be nice, but the proof eludes me. Specifically, proving that 
```
lemma list.nodup_permutations_iff {α : Type*} {l : list α} :
  nodup l.permutations ↔ nodup l :=
begin
end
```

Reply:
I tried going for the `nodup_permutations_iff` lemma. Specifically, my approach was: if on the contrary we have duplicates in `l` say at index `i`, `j`, then `l.nth \circ (equiv.swap i j)` is the same as `l.nth`, so it "just" suffices to prove that the elements of `l.permutations` are exactly the lists `l'` such that `l'.nth = l.nth \circ σ` for some permutation `σ`. But then everything turned into such a mess of yak shaving that I gave up.

So in conclusion, I don't think it's worth it to go for the clever proof in this PR.